drivers/sdhc/intel_emmc_host.c

```diff
@@ -0,0 +1,1342 @@
+{
+	volatile struct emmc_reg *regs = (struct emmc_reg *)DEVICE_MMIO_GET(dev);
+	int ret = -EAGAIN;
+	int32_t retry = time_out;
+
+	while (retry > 0) {
+		if (regs->normal_int_stat & EMMC_HOST_CMD_COMPLETE) {
+			regs->normal_int_stat = EMMC_HOST_CMD_COMPLETE;
+			ret = 0;
+			break;
+		}
+
+		k_busy_wait(1000u);
+		retry--;
+	}
+
+	if (regs->err_int_stat) {
+		LOG_ERR("err_int_stat:%x", regs->err_int_stat);
+		regs->err_int_stat &= regs->err_int_stat;
+		ret = -EIO;
+	}
+
+	if (IS_ENABLED(CONFIG_INTEL_EMMC_HOST_ADMA)) {
+		if (regs->adma_err_stat) {
+			LOG_ERR("adma error: %x", regs->adma_err_stat);
+			ret = -EIO;
+		}
+	}
+	return ret;
+}
+
+void emmc_host_sw_reset(const struct device *dev, enum emmc_sw_reset reset)
+{
+	volatile struct emmc_reg *regs = (struct emmc_reg *)DEVICE_MMIO_GET(dev);
+
+	if (reset == EMMC_HOST_SW_RESET_DATA_LINE) {
+		regs->sw_reset = EMMC_HOST_SW_RESET_REG_DATA;
+	} else if (reset == EMMC_HOST_SW_RESET_CMD_LINE) {
+		regs->sw_reset = EMMC_HOST_SW_RESET_REG_CMD;
+	} else if (reset == EMMC_HOST_SW_RESET_ALL) {
+		regs->sw_reset = EMMC_HOST_SW_RESET_REG_ALL;
+	}
+
+	while (regs->sw_reset != 0) {
+		;
+	}
+
+	k_sleep(K_MSEC(100u));
+}
+
+static int emmc_dma_init(const struct device *dev, struct sdhc_data *data, bool read)
+{
+	struct emmc_data *emmc = dev->data;
+	volatile struct emmc_reg *regs = (struct emmc_reg *)DEVICE_MMIO_GET(dev);
+
+	if (IS_ENABLED(CONFIG_DCACHE) && !read) {
+		sys_cache_data_flush_range(data->data, (data->blocks * data->block_size));
+	}
+
+	if (IS_ENABLED(CONFIG_INTEL_EMMC_HOST_ADMA)) {
+		uint8_t *buff = data->data;
+
+		/* Setup DMA trasnfer using ADMA2 */
+		memset(emmc->desc_table, 0, sizeof(emmc->desc_table));
+
+		__ASSERT_NO_MSG(data->blocks < CONFIG_INTEL_EMMC_HOST_ADMA_DESC_SIZE);
+
+		for (int i = 0; i < data->blocks; i++) {
+			emmc->desc_table[i] = ((uint64_t)buff) << EMMC_HOST_ADMA_BUFF_ADD_LOC;
+			emmc->desc_table[i] |= data->block_size << EMMC_HOST_ADMA_BUFF_LEN_LOC;
+
+			if (i == (data->blocks - 1u)) {
+				emmc->desc_table[i] |= EMMC_HOST_ADMA_BUFF_LINK_LAST;
+				emmc->desc_table[i] |= EMMC_HOST_ADMA_INTR_EN;
+				emmc->desc_table[i] |= EMMC_HOST_ADMA_BUFF_LAST;
+			} else {
+				emmc->desc_table[i] |= EMMC_HOST_ADMA_BUFF_LINK_NEXT;
+			}
+			emmc->desc_table[i] |= EMMC_HOST_ADMA_BUFF_VALID;
+			buff += data->block_size;
+			LOG_DBG("desc_table:%llx", emmc->desc_table[i]);
+		}
+
+		regs->adma_sys_addr1 = (uint32_t)((uintptr_t)emmc->desc_table & ADDRESS_32BIT_MASK);
+		regs->adma_sys_addr2 =
+			(uint32_t)(((uintptr_t)emmc->desc_table >> 32) & ADDRESS_32BIT_MASK);
+
+		LOG_DBG("adma: %llx %x %p", emmc->desc_table[0], regs->adma_sys_addr1,
+			emmc->desc_table);
+	} else {
+		/* Setup DMA trasnfer using SDMA */
+		regs->sdma_sysaddr = (uint32_t)((uintptr_t)data->data);
+		LOG_DBG("sdma_sysaddr: %x", regs->sdma_sysaddr);
+	}
+	return 0;
+}
+
+static int32_t emmc_init_xfr(const struct device *dev, struct sdhc_data *data, bool read)
```

Comment:
Why is the return type `int32_t` and not `int` here? You're assigning the return value to a variable with type `int`.